cli/cmd/awsiam.go

```diff
@@ -0,0 +1,325 @@
+//
+// Author:: Nicholas Schmeller (<nick.schmeller@lacework.net>)
+// Copyright:: Copyright 2023, Lacework Inc.
+// License:: Apache License, Version 2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/iam"
+	"github.com/aws/aws-sdk-go-v2/service/iam/types"
+	"github.com/lacework/go-sdk/lwrunner"
+)
+
+// setupSSMAccess sets up an IAM role for SSM and attaches it to
+// the machine's instance profile. Takes role name as argument;
+// pass the empty string to create a new role.
+// Then creates SSM document.
+func setupSSMAccess(cfg aws.Config, roleName string, token string) (types.Role, types.InstanceProfile, error) {
+	cli.Log.Debugw("setting up role", "passed roleName", roleName)
+	role, err := setupSSMRole(cfg, roleName)
+	if err != nil {
+		return role, types.InstanceProfile{}, err
+	}
+
+	err = attachSSMPoliciesToRole(cfg, role)
+	if err != nil {
+		return role, types.InstanceProfile{}, err
+	}
+
+	// Create instance profile and add the role to it
+	instanceProfile, err := setupInstanceProfile(cfg, role)
+	if err != nil {
+		return role, instanceProfile, err
+	}
+
+	return role, instanceProfile, nil
+}
+
+// teardownSSMAccess destroys all the infra created during the execution of this program.
+// Specifically, this function:
+// - Removes the role from the instance profile
+// - Deletes the instance profile
+// - Detaches all managed policies from the role (assumes no inline policies attached)
+// - Deletes the role
+func teardownSSMAccess(cfg aws.Config, role types.Role, instanceProfile types.InstanceProfile, byoRoleName string) error {
+	c := iam.New(iam.Options{
+		Credentials: cfg.Credentials,
+		Region:      cfg.Region,
+	})
+
+	cli.Log.Debug("only destroying instance profile if it's ours")
+	if *instanceProfile.InstanceProfileName == instanceProfileName {
```

Comment:
I think that this will be safer if you examine a tag created by this script rather than the role name, I could imagine refactorings very easily forgetting that we care specifically about the role name and accidentally deleting a customer's instance profile.

Reply:
I agree, plus tags seem a bit more kosher. Done with 162cf4a1ca3a9722a25341408478d7f6edaf476c